hw/ip/pwm/doc/_index.md

```diff
@@ -0,0 +1,181 @@
+In some applications, the user may wish to stagger outputs to limit the overall magnitude of current spikes in the complete system.
+Thus each output also has a 16-bit phase control register.
+In security applications it should be noted, however, that it is the system-designer's responsibility to ensure that improper configuration of the phase control registers would not pose a security risk to the overall system (due to, for instance, a denial of service attack through the PWM configuration).
+
+The PWM IP also provides a hardware-controlled blink feature, which can periodically drive a PWM output to zero for many PWM cycles.
+The blink-rate for each channel is independently defined relative to the common PWM pulse-frequency via a pair of 8-bit clock divider and clock pre-divider registers.
+The duty-cycle and phase of the blink rate can also be controlled, each with up to 8-bit resolution.
+Each channel also has a separate blink-enable register bit, allowing for multiple outputs to be configured to blink synchronously.
+
+# Theory of Operations
+
+## Hardware Interfaces
+
+{{< hwcfg "hw/ip/pwm/data/pwm.hjson" >}}
+
+## Design Details
+
+The IP consists of a single 16-bit phase counter and a number of PWM FSM instances, one per output.
+
+### PWM Phase Counter
+
+Each PWM pulse cycle is divided into 2<sup>16</sup> beats.
+During each beat, the 16-bit phase counter increments by one (modulo 65536).
+The beat period is defined by the {{< regref "CFG.CLK_DIV" >}} register:
+$$f_\textrm{beat}=\frac{f_\textrm{I/O clk}}{\textrm{CLK_DIV}+1}$$
+
+A PWM pulse cycle is completed each time the phase counter overflows to 0.
+The PWM drive frequency is therefore:
+$$f_\textrm{PWM}=\frac{f_\textrm{I/O clk}}{2^{16}(\textrm{CLK_DIV}+1)}$$
+
+The PWM phase counter is reset whenever {{< regref "CFG.CNTR_EN" >}} is disabled.
+
+The following figure illustrates the effect of the clock divider register.  Note that changes to {{< regref "CFG.CLK_DIV" >}} only take effect when {{< regref "CFG.CNTR_EN" >}} is disabled.
+
+{{< wavejson >}}
+{signal: [
+  {name: 'clk', wave: 'p..............|.........'},
+  {name: 'PWM_EN', wave: '0..1...........|01.......'},
+  {name: 'CLK_DIV', wave: '2....4.........|.2.......', data: '2 4 4'},
+  {name: 'clk_div_int', wave: '2................2.......', data: '2 4'},
+  {name: 'phase_ctr', wave: '3..2..2..2..2..|32....2..', data: '0 0 1 2 3 0 0 1'},
+  {name: 'clk_div_ctr', wave: '2...22222222222|2.2222222', data: '0 1 2 0 1 2 0 1 2 0 1 2 0 1 2 3 4 0 1 2'},
+  {name: 'clk_div_pulse', wave: '0....10.10.10.1|0....10..'}],
+  config:{skin:'narrow'}
+}
+{{< /wavejson >}}
+
+### PWM FSMs
+
+Each output is driven by an independent FSM.
+Each takes the phase-counter as a common input, as well as the following channel-specific register inputs:
+- a single PWM enable bit (e.g. {{< regref "PWM_EN.EN_0" >}})
+- a 16-bit PWM duty cycle (e.g. {{< regref "PWM_PARAM_0.PHASE_DELAY_0" >}})
```

Comment:
looks like the regrefs are reversed for this item and the one below.

Reply:
Oh.  Thanks, good catch.

Reply:
(This section has been completely rewritten)